src/metabase/models/dashboard_card.clj

```diff
@@ -301,30 +304,31 @@
      :from     [[{:union-all (map link-card-info-query-for-model link-card-model->ids)}
                  :alias_is_required_by_sql_but_not_needed_here]]}))
 
-(defn with-link-card-info
+(mi/define-batched-hydration-method dashcard-linkcard-info
+  :dashcard/linkcard-info
   "Update entity info for link cards.
 
   Link cards are dashcards that link to internal entities like Database/Dashboard/... or an url.
   The viz-settings only store the model name and id, info like name, description will need to be
   hydrated on fetch to make sure those info are up-to-date."
   [dashcards]
-  (let [entity-path        [:visualization_settings :link :entity]
+  (let [entity-path   [:visualization_settings :link :entity]
         ;; find all dashcards that are link-cards and get its model, id
         ;; [[:table #{1 2}] [:database #{3 4}]]
-        model-and-ids      (->> dashcards
-                                (filter #(link-card-models (keyword (get-in % (conj entity-path :model)))))
-                                (map #(get-in % entity-path))
-                                (group-by :model)
-                                (map (fn [[k v]] [(keyword k) (set (map :id v))])))
-        ;; query all entities in 1 db call
-        ;; {[:table 3] {:name ...}}
-        model-and-id->info (when (seq model-and-ids)
-                             (-> (m/index-by (juxt :model :id) (t2/query (link-card-info-query model-and-ids)))
-                                 (update-vals (fn [{model :model :as instance}]
-                                                (if (mi/can-read? (t2/instance (link-card-model->toucan-model (keyword model)) instance))
-                                                  instance
-                                                  {:restricted true})))))]
-    (if model-and-id->info
+        model-and-ids (->> dashcards
+                           (map #(get-in % entity-path))
+                           (filter #(link-card-models (:model %)))
+                           (group-by :model)
+                           (map (fn [[k v]] [k (set (map :id v))])))]
+    (if-let [model-and-id->info
+             ;; query all entities in 1 db call
+             ;; {[:table 3] {:name ...}}
+             (and model-and-ids
```

Comment:
I still see two conditionals here but you swapped `when` with `and` and `if` with `if-let`.

My suggestion was to use `(if (seq model-and-ids) (let [model-and-id->info ...] ...) dashcards)` to only need one conditional.